src/cargo/sources/path.rs

```diff
@@ -411,6 +465,153 @@ impl<'gctx> PathSource<'gctx> {
         }
     }
 
+    /// Lists files relevant to building this package inside this source by
+    /// traversing the git working tree, while avoiding ignored files.
+    ///
+    /// This looks into Git sub-repositories as well, resolving them to individual files.
+    /// Symlinks to directories will also be resolved, but walked as repositories if they
+    /// point to one to avoid picking up `.git` directories.
+    fn list_files_gix(
+        &self,
+        pkg: &Package,
+        repo: &gix::Repository,
+        filter: &dyn Fn(&Path, bool) -> bool,
+    ) -> CargoResult<Vec<PathBuf>> {
+        warn!("list_files_gix {}", pkg.package_id());
+        let options = repo
+            .dirwalk_options()?
+            .emit_untracked(gix::dir::walk::EmissionMode::Matching)
+            .emit_ignored(None)
+            .emit_tracked(true)
+            .recurse_repositories(false)
+            .symlinks_to_directories_are_ignored_like_directories(true)
+            .emit_empty_directories(false);
+        let index = repo.index_or_empty()?;
+        let root = repo
+            .work_dir()
+            .ok_or_else(|| anyhow::format_err!("can't list files on a bare repository"))?;
+        assert!(
+            root.is_absolute(),
+            "BUG: paths used internally are absolute, and the repo inherits that"
+        );
+
+        let pkg_path = pkg.root();
+        let mut target_prefix;
+        let repo_relative_pkg_path = pkg_path.strip_prefix(root).unwrap_or(Path::new(""));
+
+        let pathspec = {
+            let mut include = gix::path::to_unix_separators_on_windows(gix::path::into_bstr(
+                repo_relative_pkg_path,
+            ));
+
+            target_prefix = include.clone();
+            target_prefix.to_mut().push_str(if include.is_empty() {
+                "target/"
+            } else {
+                "/target/"
+            });
+
+            // The `target` directory is never included if it is in the package root.
+            let exclude = {
+                let mut pattern = include.clone();
+                pattern
+                    .to_mut()
+                    .insert_str(0, if include.is_empty() { ":!" } else { ":!/" });
+                pattern.to_mut().push_str(b"/target/");
+                pattern
+            };
+            include.to_mut().insert_str(0, ":/");
+
+            vec![include.into_owned(), exclude.into_owned()]
+        };
```

Comment:
I've attempted to simplify the pathspec code to make it easier to understand. If this is still correct, then I think it's easier to follow (at least for me).
 
```suggestion
        let pkg_path = pkg.root();
        let repo_relative_pkg_path = pkg_path.strip_prefix(root).unwrap_or(Path::new(""));
        let target_prefix = gix::path::to_unix_separators_on_windows(gix::path::into_bstr(
            repo_relative_pkg_path.join("target"),
        ));
        let package_prefix =
            gix::path::to_unix_separators_on_windows(gix::path::into_bstr(repo_relative_pkg_path));

        let pathspec = {
            // Include the package root.
            let mut include = BString::from(":/");
            include.push_str(package_prefix.as_ref());

            // Exclude the target directory.
            let mut exclude = BString::from(":!");
            exclude.push_str(target_prefix.as_ref());

            vec![include, exclude]
        };
```

Reply:
Thanks! Much easier to read!

I will commit the suggestion and then work on assuring it's truly the same, I *think* there are slash-related issues right now that alter the meaning of the pathspecs.

Reply:
Thanks a lot for your help with this!

Leveraging `Path::join` and building BStrings front to back really make all the difference here.